### PR TITLE
Deadlines

### PR DESCRIPTION
**Pull request by akaspin**
[trello](https://trello.com/c/0V5so0Jr/291-go-server-tls-handshake-can-hangs-indefinitely)

TLS `Handshake()` and  `go-server` Close()` (in some cases) means *writes*. In case of "bad" or "dead" connections deadline will prevent zombie connections and goroutines.

### CODE REVIEW

**Comment by dynamix** [comment]
comment in trello https://trello.com/c/pRRyQsYk/3472-go-server-avoid-potential-oom-issues

**Comment by akaspin** [comment]
@dynamix I agree that pause is bad idea (changes pushed). But deadlines are still in game.

**Comment by akaspin** [comment]
reopening after discussion

**Comment by akaspin** [comment]
[[go-server] TLS handshake can hangs indefinitely](https://trello.com/c/0V5so0Jr/291-go-server-tls-handshake-can-hangs-indefinitely)

File: server.go [hunk 1]
```diff
@@ -152,19 +155,24 @@ func (server *Server) serve(listener *Listener) error {
...
+		// for cases of probe or blackhole connection
+		if err := conn.SetDeadline(time.Now().Add(connTimeout)); err != nil {
+			server.Log.Warnf("can not set deadline for conn: %v", err)
+			continue
+		}
 
-		if server.MaxConns > 0 && server.numConns.Count() >= server.MaxConns {
-			// too many connections
+		if server.MaxConns > 0 && server.numConns.Count() > server.MaxConns {
+			// temporary pause accept loop
```
Comment (dynamix) [inline server.go]: remove invalid comment

File: connection.go [hunk 2]
```diff
@@ -142,15 +149,21 @@ func (c *Connection) Close() {
 		c.Server.closes.Inc(1)
 	}
 
-	// flush write buffer before close
-	if c.Buffer.Writer != nil {
-		c.Buffer.Writer.Flush()
+	if c.Conn != nil {
+		// set guard deadline in case of bad connection
+		if err := c.Conn.SetDeadline(time.Now().Add(connTimeout)); err != nil {
+			c.Server.Log.Warnf("can not set conn deadline before close: %v", err)
```
Comment (dynamix) [inline connection.go]: remove unnecessary logging + error handling

File: connection.go [hunk 3]
```diff
@@ -124,6 +125,12 @@ func (c *Connection) Serve() {
 		c.Server.numHandshakes.Dec(1)
 	}
 
+	// reset deadline before handle
+	if err := c.Conn.SetDeadline(time.Time{}); err != nil {
+		c.Server.Log.Warnf("can not reset conn deadline: %v", err)
```
Comment (dynamix) [inline connection.go]: Remove unnecessary logging

File: server.go [hunk 4]
```diff
@@ -3,11 +3,14 @@ package server
 import (
 	"crypto/tls"
 	"fmt"
+	"time"
 
 	"github.com/rcrowley/go-metrics"
 	"github.com/remerge/cue"
 )
 
+const connTimeout = 100 * time.Millisecond
```
Comment (dynamix) [inline server.go]: This needs to be configurable + 100ms seems way to low as a default.

File: server.go [hunk 5]
```diff
@@ -152,19 +155,24 @@ func (server *Server) serve(listener *Listener) error {
 			return err
 		}
 
-		server.accepts.Inc(1)
+		// for cases of probe or blackhole connection
+		if err := conn.SetDeadline(time.Now().Add(connTimeout)); err != nil {
+			server.Log.Warnf("can not set deadline for conn: %v", err)
```
Comment (dynamix) [inline server.go]: unnecessary logging

**Review by dynamix — CHANGES_REQUESTED** [review]
makes sense but lots of unnecessary logging

File: server.go [hunk 6]
```diff
@@ -152,19 +155,22 @@ func (server *Server) acceptLoop(listener *Listener) error {
...
-		if server.MaxConns > 0 && server.numConns.Count() >= server.MaxConns {
-			// too many connections
+		if server.MaxConns > 0 && server.numConns.Count() > server.MaxConns {
 			server.tooManyConns.Inc(1)
-			conn.Close()
+			_ = conn.Close()
 			continue
 		}
 
+		server.accepts.Inc(1)
```
Comment (dynamix) [inline server.go]: shouldn't this happen directly after `listener.Accept ` returned without an error? As it count the number of accepts before we close connection incase of too many open connections?

File: server.go [hunk 7]
```diff
@@ -152,19 +155,22 @@ func (server *Server) acceptLoop(listener *Listener) error {
 			return err
 		}
 
-		server.accepts.Inc(1)
+		// for cases of probe or blackhole connection
+		if err := conn.SetDeadline(time.Now().Add(server.Timeout)); err != nil {
+			continue
```
Comment (dynamix) [inline server.go:193]: If this fails, shouldn't we `Close()` the connection as well?

File: server.go [hunk 8]
```diff
@@ -33,11 +35,12 @@ type Server struct {
...
-	server = &Server{}
-
-	server.Id = fmt.Sprintf("server:%d", port)
-	server.Port = port
-	server.BufferSize = 32768
+	server = &Server{
+		Id:         fmt.Sprintf("server:%d", port),
+		Port:       port,
+		BufferSize: 32768,
+		Timeout:    500 * time.Millisecond,
```
Comment (dynamix) [inline server.go]: This could be a bit higher to be on the safe side.

**Review by dynamix — CHANGES_REQUESTED** [review]
Looks good, just some minor comments

File: server.go (hunk 8 above)
Comment (akaspin) [reply server.go]: This value affects sets deadlines *only* for TLS. Let's change to three secs.

File: connection.go [hunk 9]
```diff
@@ -142,15 +148,20 @@ func (c *Connection) Close() {
 		c.Server.closes.Inc(1)
 	}
 
-	// flush write buffer before close
-	if c.Buffer.Writer != nil {
-		c.Buffer.Writer.Flush()
+	if c.Conn != nil {
+		// set guard deadline in case of bad connection
+		if err := c.Conn.SetDeadline(time.Now().Add(c.Server.Timeout)); err != nil {
+			return
```
Comment (gburanov) [inline connection.go:155]: should not it be still closed even if setting deadline does not work?

File: server.go [hunk 10]
```diff
@@ -17,6 +17,7 @@ type Server struct {
 	MaxConns                   int64
 	MaxConcurrentTLSHandshakes int64
 	BufferSize                 int
+	Timeout                    time.Duration // Service timeout. Default is 500ms
```
Comment (gburanov) [inline server.go]: "Default is 500ms" does not match "Timeout:    3 * time.Second,". Maybe we can remove whole comment.

**Review by gburanov — CHANGES_REQUESTED** [review]

**Review by gburanov — APPROVED** [review]

**Pull request merged**